content/2018/12/10/channels-and-drf.rst

```diff
@@ -0,0 +1,292 @@
+
+            We then need to use that validated data to confirm that the
+            requesting user (available in self.scope["user"] because of
+            the use of channels.auth.AuthMiddlewareStack in routing) is
+            allowed to subscribe to the requested object.
+            """
+
+            # Define this method on your consumer like the version in
+            # rest_framework.generics.GenericAPIView
+            # Be sure to pass in the context, so that the serializer can
+            # confirm that this particular user can get updates on this
+            # particular object.
+            serializer = self.get_serializer(data=content)
+            if not serializer.is_valid():
+                return
+            # Define this method on your serializer:
+            group_name = serializer.get_group_name()
+            # The AsyncJsonWebsocketConsumer parent class has a
+            # self.groups list already. It uses it in cleanup.
+            self.groups.append(group_name)
+            # This actually subscribes the requesting socket to the
+            # named group:
+            await self.channel_layer.group_add(
+                group_name,
+                self.channel_name,
+            )
+
+And you'll want to add some stuff to your ``routing`` module, too::
+
+    from django.urls import path
+
+    from channels.auth import AuthMiddlewareStack
+    from channels.routing import ProtocolTypeRouter, URLRouter
+
+    from .consumers import NotificationConsumer
+
+
+    websockets = URLRouter([
+        path(
+            "ws/notifications/",
+            NotificationConsumer,
+            name="ws_notifications",
+        ),
+    ])
+
+
+    application = ProtocolTypeRouter({
+        # (http->django views is added by default)
+        "websocket": AuthMiddlewareStack(websockets),
+    })
+
+There are a couple more pieces. We need to actually send updates when a
+model changes!
+
+We separate out those concerns. We add a ``notifications`` module with
+the appropriate functions to wrap up the data and send it over the
+channels layer, and then we call out to those functions in the models'
+``save`` methods.
+
+In our ``notifications.py`` we have something like this::
+
+
+    from channels.layers import get_channel_layer
+    from .serializers import FooSerializer
+
+    async def update_foo(instance):
+        serializer = FooSerializer(foo)
+        group_name = serializer.get_group_name()
+        channel_layer = get_channel_layer()
+        content = {
+            # This "type" passes through to the frontend to facilitate
+            # our Redux events.
+            "type": "UPDATE_FOO",
+            "payload": serializer.data,
+        }
+        await channel_layer.group_send(group_name, {
+            # This "type" defines which handler on the Consumer gets
+            # called.
+            "type": "notify",
+            "content": content,
+        })
+
+And then our models has something like this::
```

Comment:
have over has here? But also, we ended the last two lines with "... something like this" so maybe it is better to say something along the lines of "Then our models end up looking like:"